src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlySpanND.cs

```diff
@@ -0,0 +1,468 @@
+        /// Thrown when <typeparamref name="T"/> is reference type or contains pointers and hence cannot be stored in unmanaged memory.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the specified length is negative.
+        /// </exception>
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe ReadOnlySpanND(void* pointer, ReadOnlySpan<nint> lengths, bool isPinned, ReadOnlySpan<nint> strides = default)
+        {
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
+
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
+            if (_linearLength < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+            if (_linearLength != SpanNDHelpers.CalculateTotalLength(lengths))
+                ThrowHelper.ThrowArgument_LengthsMustEqualArrayLength();
+
+            _isPinned = isPinned;
+
+#pragma warning disable CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
+            _reference = ref *(T*)pointer;
+#pragma warning restore CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
+
+            _lengths = lengths.ToArray();
+            _strides = strides.ToArray();
+            if (strides == ReadOnlySpan<nint>.Empty)
+                _strides = SpanNDHelpers.CalculateStrides(Rank, lengths);
+        }
+
+        // Constructor for internal use only. It is not safe to expose publicly, and is instead exposed via the unsafe MemoryMarshal.CreateReadOnlySpan.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlySpanND(ref T reference, ReadOnlySpan<nint> lengths, ReadOnlySpan<nint> strides, bool isPinned)
+        {
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
+            Debug.Assert(_linearLength >= 0);
+
+            _reference = ref reference;
+
+            _lengths = lengths.ToArray();
+            _strides = strides.ToArray();
+            _isPinned = isPinned;
+        }
+
+        /// <summary>
+        /// Returns a reference to specified element of the ReadOnlySpanND.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref readonly T this[params ReadOnlySpan<nint> indices]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (indices.Length != Rank)
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+
+                var index = SpanNDHelpers.GetIndex(indices, Strides, Shape);
```

Comment:
nit: The libraries repo only uses `var` when the type is "apparent", which this doesn't qualify as.

(I also prefer to use var almost everywhere, but we need to stay consistent with the repo style guidelines).

Reply:
Yeah, In the process of converting all `var` back to the explicit type.